src/java.base/share/classes/java/util/zip/GZIPOutputStream.java

```diff
@@ -36,43 +36,58 @@
  *
  */
 public class GZIPOutputStream extends DeflaterOutputStream {
+
     /**
      * CRC-32 of uncompressed data.
      */
-    protected CRC32 crc = new CRC32();
-
-    /*
-     * GZIP header magic number.
-     */
-    private static final int GZIP_MAGIC = 0x8b1f;
+    private CRC32 crc = new CRC32();
 
-    /*
+    /**
      * Trailer size in bytes.
      *
      */
     private static final int TRAILER_SIZE = 8;
 
-    // Represents the default "unknown" value for OS header, per RFC-1952
-    private static final byte OS_UNKNOWN = (byte) 255;
-
     /**
-     * Creates a new output stream with the specified buffer size.
-     *
-     * <p>The new output stream instance is created as if by invoking
-     * the 3-argument constructor GZIPOutputStream(out, size, false).
+     * Creates a new output stream with the specified buffer size,
+     * flush mode flags and header fields.
      *
      * @param out the output stream
      * @param size the output buffer size
+     * @param syncFlush
+     *        if {@code true} invocation of the inherited
+     *        {@link DeflaterOutputStream#flush() flush()} method of
+     *        this instance flushes the compressor with flush mode
+     *        {@link Deflater#SYNC_FLUSH} before flushing the output
+     *        stream, otherwise only flushes the output stream
+     * @param gzipHeaderData
+     *        The header of Gzip file, contains header members defined
+     *        in RFC 1952. if {@code null}, use default header data.
+     *
      * @throws    IOException If an I/O error has occurred.
-     * @throws    IllegalArgumentException if {@code size <= 0}
+     *
+     * @since 18
      */
-    public GZIPOutputStream(OutputStream out, int size) throws IOException {
-        this(out, size, false);
+    public GZIPOutputStream(OutputStream out,
+                            int size,
+                            boolean syncFlush,
+                            GZIPHeaderBuilder.GZIPHeaderData gzipHeaderData)
```

Comment:
Given the Gzip header data is so infrequently modified,  one additional constructor is all that really need.